autogo/shadow.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type visitor struct {
+	origin string
+	dest   string
+}
+
+func (v visitor) visit(path string, info os.FileInfo, verr error) error {
+	r, err := filepath.Rel(v.origin, path)
+	if err != nil {
+		return fmt.Errorf("%q is not relative to %q: %v", path, v.origin, err)
+	}
+	if r == ".." || strings.HasPrefix(r, "../") {
+		return fmt.Errorf("%q is not a child of %q", path, v.origin)
+	}
+	if strings.HasSuffix(path, "/testdata") || strings.Contains(path, "/testdata/") {
```

Comment:
technically I don't think that's correct, since it's (unfortunately) valid to import from a testdata dir.
See https://golang.org/pkg/go/build/testdata/other/file for example
Maybe at least a TODO/comment?